virtcontainers/vm.go

```diff
@@ -0,0 +1,233 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &VM{
+		id:         id,
+		hypervisor: hypervisor,
+		agent:      agent,
+		cpu:        config.HypervisorConfig.DefaultVCPUs,
+		memory:     config.HypervisorConfig.DefaultMemSz,
+	}, nil
+}
+
+func buildVMSharePath(id string) string {
+	return filepath.Join(RunVMStoragePath, id, "shared")
+}
+
+func (v *VM) logger() logrus.FieldLogger {
+	return virtLog.WithField("vm id", v.id)
+}
+
+// Pause pauses a VM.
+func (v *VM) Pause() error {
+	v.logger().Info("pause vm")
+	return v.hypervisor.pauseSandbox()
+}
+
+// Save saves a VM to persistent disk.
+func (v *VM) Save() error {
+	v.logger().Info("save vm")
+	return v.hypervisor.saveSandbox()
+}
+
+// Resume resumes a paused VM.
+func (v *VM) Resume() error {
+	v.logger().Info("resume vm")
+	return v.hypervisor.resumeSandbox()
+}
+
+// Start kicks off a configured VM.
+func (v *VM) Start() error {
+	v.logger().Info("start vm")
+	return v.hypervisor.startSandbox()
+}
+
+// Kill stops a VM process.
+func (v *VM) Kill() error {
+	v.logger().Info("kill vm")
+	return v.hypervisor.stopSandbox()
+}
+
+// AddCPUs adds num of CPUs to the VM.
+func (v *VM) AddCPUs(num uint32) error {
+	if num > 0 {
+		v.logger().Info("hot adding %d vCPUs", num)
+		if _, err := v.hypervisor.hotplugAddDevice(num, cpuDev); err != nil {
+			return err
+		}
+		v.cpuDelta += num
+		v.cpu += num
+	}
+
+	return nil
+}
+
+// AddMemory adds numMB of memory to the VM.
+func (v *VM) AddMemory(numMB uint32) error {
+	if numMB > 0 {
+		v.logger().Info("hot adding %d MB memory", numMB)
+		dev := &memoryDevice{1, int(numMB)}
+		if _, err := v.hypervisor.hotplugAddDevice(dev, memoryDev); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// OnlineCPUMemory puts the hotplugged CPU and memory online.
+func (v *VM) OnlineCPUMemory() error {
+	v.logger().Info("online CPU %d and memory", v.cpuDelta)
+	err := v.agent.onlineCPUMem(v.cpuDelta)
+	if err == nil {
+		v.cpuDelta = 0
+	}
+
+	return err
+}
+
+func (v *VM) assignSandbox(s *Sandbox) error {
+	// add vm symlinks
+	// - link vm socket from sandbox dir (/run/vc/vm/sbid/<kata.sock>) to vm dir (/run/vc/vm/vmid/<kata.sock>)
+	// - link 9pfs share path from sandbox dir (/run/kata-containers/shared/sandboxes/sbid/) to vm dir (/run/vc/vm/vmid/shared/)
+
+	vmSharePath := buildVMSharePath(v.id)
+	vmSockDir := v.agent.getVMPath(v.id)
+	sbSharePath := s.agent.getSharePath(s.id)
+	sbSockDir := s.agent.getVMPath(s.id)
+
+	v.logger().WithFields(logrus.Fields{
+		"vmSharePath": vmSharePath,
+		"vmSockDir":   vmSockDir,
+		"sbSharePath": sbSharePath,
+		"sbSockDir":   sbSockDir,
+	}).Infof("assign vm to sandbox %s", s.id)
+
+	// First make sure the symlinks do not exist
+	os.RemoveAll(sbSharePath)
+	os.RemoveAll(sbSockDir)
+
+	err := os.Symlink(vmSharePath, sbSharePath)
+	if err != nil {
+		return err
+	}
+
+	err = os.Symlink(vmSockDir, sbSockDir)
+	if err != nil {
+		os.Remove(sbSharePath)
+		return err
+	}
+
+	s.hypervisor = v.hypervisor
```

Comment:
I understand that we assign an existing hypervisor handler to the sandbox, but shouldn't we also assign the agent handler to this same sandbox ?

Reply:
Not, the vm factory uses a builtin proxy w/ `longliveconn` set to false. So the agent may not be reusable by a normal sandbox, which might use a normal proxy, and yes, as you can see, agent and proxy are tightly coupled in virtcontainers. Also since we are not using long live agent connections, there is no overhead with dropping the existing agent handler. In fact sandbox->agent is already usable by the time this is called.